### ISSUE TITLE
instable result for a thirdparty testsuite

### ISSUE THREAD

**Issue by tst2005** [pygy/LuLPeg#10]
Hello pygy,
It's me again :-D

Almost like the [peg issue](https://github.com/pygy/LuLPeg/issues/6), I checked [lua-parser](https://github.com/andremm/lua-parser) with LuLPeg : Awsome, all tests passed with lua5.1/5.2 !
...but in fact ... most of the time yes but not always.
It's was so strange, I can resist to inform you !
If you have time, take a look at https://github.com/andremm/lua-parser/issues/2

Regards,

**Comment by pygy** [comment]
Now this is most strange.

I'm currently focusing on https://github.com/j2css/j2c/ and related project, but once I'm done with j2c v1.0 I'll dive in.

**Comment by pygy** [comment]
BTW, thanks for the report, off course.

**Comment by HeinrichHartmann** [comment]
Confirming this issue.

We have seen random errors like:

```
[2016-05-27 18:36:19.978223] lua: /opt/circonus/lib/amd64/snowth/support/lulpeg.lua:613: pattern error near '{| delt (ds delt)* |}...'
```
That are non-sensical and very hard to reproduce. Most of the time parsing just works fine/

We also have seen interactions between different grammars!

E.g.
```
p = re.compile(GRAMMAR_1)
q = re.compile(GRAMMAR_2)
```
And q throws errors related to `GRAMMAR_1`. This has been a while back, and we solved it by making the rule names unique across our lua state. But this is a very bad sign, that raises alarm bells!

We suspect some global state in the lib, that causes all these problems.

**Comment by pygy** [comment]
That's eerily weird. The only global state is IIRC the pattern cache, so that `P'x' == P'x'` (which is used internally to simplify some patterns)... Maybe there's some false positive hit and a bad pattern is sometimes returned...

nother thing that can trigger delayed warnings is the fact that  grammars are only compiled to Lua code the first time their `:match()` method is called.

@HeinrichHartmann the problem occurs when calling `q = re.compile(GRAMMAR_2)`, correct?

Could you try to call (`p:match''`) before running that line?

If possible at, could you share the problematic grammars so that I can try to reproduce the issue?

**Comment by jbrophy60** [comment]
I have come across the same issue in my project. I get very occasional parsing failures which are not deterministic, as executing again with the same data does not reproduce the failure.
I have been experimenting with the tests from tst2005 and can reproduce these failures fairly easily, although again, not deterministically.
I discovered that immediately re-parsing the failed grammar always gives the same failure but forcing a reload of lulpeg.lua allows the test to pass.
I have also experimented with making lua more deterministic by providing a custom version of next() and regular calls to collectgarbage(). This seems to allow the tests to pass every time.
I hope this provides some pointers to a solution.

**Comment by pygy** [comment]
Thanks for the analysis @jbrophy60 I'll try to push a fix soon.

**Comment by Totktonada** [comment]
@pygy Do you have any updates? We meet something like on our own test suite (in a project that uses lulpeg) after some changes that are seems to be completely unrelated. Don't sure how to proceed with that.

**Comment by tst2005** [comment]
I'm really happy to see this issue closed.
Thanks guys for all the work done!

**Comment by tst2005** [comment]
@pygy is it planned to release a new lulpeg version including this fix (maybe v0.1.1) ?

**Comment by pygy** [comment]
@tst @Totktonada v0.1.1 it is. I'm squeezed right now, so I won't release it immediately.
